### PR TITLE
CLI drive selection

### PR DESCRIPTION
**Pull request by stefan-mihaila**
Still need to write some unit tests, but I wanted to open the PR so I can get some feedback.

screen shot 2016-12-14 at 16 49 08
screen shot 2016-12-14 at 16 49 21

### CODE REVIEW

File: lib/cli/etcher.js [hunk 1]
```diff
@@ -18,62 +18,114 @@
...
+  }
+  return choice;
+};
+
+const mustConfirm = !options.robot && !options.yes;
+
+Bluebird.all([
+  isElevated(),
+  imageStream.getImageMetadata(options._[0])
+]).then((results) => {
```
Comment (stefan-mihaila) [inline lib/cli/etcher.js]: I would've loved to use rest parameters destruction here `then(...[elevated, imageMetadata')`, but `ESLint` complains.

File: npm-shrinkwrap.json [hunk 2]
```diff
@@ -101,6 +101,23 @@
       "from": "any-promise@>=1.1.0 <2.0.0",
       "resolved": "https://registry.npmjs.org/any-promise/-/any-promise-1.3.0.tgz"
     },
+    "appdmg": {
```
Comment (stefan-mihaila) [inline npm-shrinkwrap.json]: I don't know why this changes are here, I just added `inquirer` and removed `resin-cli-form` in `package.json`. @jviotti can you tell me if it should be this way?

Comment (jviotti) [inline npm-shrinkwrap.json]: Did you run `npm run shrinkwrap` after installing the dependencies? That script is exactly meant to remove these kind of things.

File: npm-shrinkwrap.json [hunk 3]
```diff
@@ -1792,6 +1826,11 @@
       "from": "fs-temp@>=1.0.0 <2.0.0",
       "resolved": "https://registry.npmjs.org/fs-temp/-/fs-temp-1.1.0.tgz"
     },
+    "fs-xattr": {
```
Comment (jviotti) [inline npm-shrinkwrap.json]: This one should have been removed by `npm run shrinkwrap` as well.

File: lib/cli/etcher.js [hunk 4]
```diff
@@ -18,61 +18,115 @@
...
-      device: answers.drive
+      device: drive
     });
 
     if (!selectedDrive) {
-      throw new Error(`Drive not found: ${answers.drive}`);
+      throw new Error(`Drive not found: ${drive}`);
+    }
+
+    if (selectedDrive.system && mustConfirm) {
```
Comment (jviotti) [inline lib/cli/etcher.js]: This should re-use logic from `SelectionStateModel`.

File: lib/cli/etcher.js [hunk 5]
```diff
@@ -18,61 +18,115 @@
...
+      return inquirer.prompt([ {
```
Comment (jviotti) [inline lib/cli/etcher.js]: Lets try to refactor this so there is a single `inquirer.prompt()` invocation in `etcher.js`. Conditional questions can be made with `when()`.

**Comment by lurch** [comment]
I realize that things are currently in flux ( #888 ), but for this initial version of the PR I wonder if the "labels" (`FIXED DRIVE`, `NON-REMOVABLE DRIVE`, etc.) should be matched up with what the GUI displays?

File: lib/cli/drive-scanner.js [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+
+'use strict';
+
+const drivelist = require('drivelist');
+const EventEmitter = require('events').EventEmitter;
+const _ = require('lodash');
+const Rx = require('rx');
+
+const interval = new WeakMap();
+const prevPollDrives = new WeakMap();
```
Comment (jviotti) [inline lib/cli/drive-scanner.js]: What is the reason to use maps if we only get/set from/to a single key (`this`)?

File: lib/cli/dynamic-list.js [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+const inquirer = require('inquirer');
+const Choices = require('inquirer/lib/objects/choices');
+
+const updateChoices = function(choices) {
+  this.opt.choices = new Choices(choices);
+  this.render();
+};
+
+/**
+ * Prompt for `inquirer` that adds dynamic update functionality to the `list` prompt.
```
Comment (jviotti) [inline lib/cli/dynamic-list.js]: This should be prefixed with `@summary`.

File: lib/cli/dynamic-list.js [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+ *
+ * @class
+ * @public
+ * @extends inquirer.prompt.prompts.list
+ *
+ * @example
+ *
```
Comment (jviotti) [inline lib/cli/dynamic-list.js]: Any reason for this while line? If we decide to process the JSDoc from this file, there would be a weird blank line on top.

File: lib/cli/dynamic-list.js [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+ * inquirer.registerPrompt('dynamicList', DynamicList);
+ *
+ * inquirer.prompt([{
+ *   message: ''What size do you need?'',
+ *   type: 'dynamicList',
+ *   choices: ['Jumbo', 'Large', 'Standard', 'Medium', 'Small', 'Micro'],
+ *   emitter: scanner,  // Event emitter with events `add` and `remove` and methods `start` and `stop`
+ *   name: 'size',
+ *   mapFn: _.identity  // mapFn is used transform one choice from the emitter into a `list` choice
+ *  });
```
Comment (jviotti) [inline lib/cli/dynamic-list.js]: The indentation here is a bit off.

File: lib/cli/dynamic-list.js [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+ * @class
+ * @public
+ * @extends inquirer.prompt.prompts.list
+ *
+ * @example
+ *
+ * inquirer.registerPrompt('dynamicList', DynamicList);
+ *
+ * inquirer.prompt([{
+ *   message: ''What size do you need?'',
```
Comment (jviotti) [inline lib/cli/dynamic-list.js]: There are double single quotes here.

File: lib/cli/dynamic-list.js [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+ *   type: 'dynamicList',
+ *   choices: ['Jumbo', 'Large', 'Standard', 'Medium', 'Small', 'Micro'],
+ *   emitter: scanner,  // Event emitter with events `add` and `remove` and methods `start` and `stop`
```
Comment (jviotti) [inline lib/cli/dynamic-list.js]: Here is says that the scanner should emit `add` and `remove`, but the current implementation emits `change`.

File: lib/cli/dynamic-list.js [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+ * @extends inquirer.prompt.prompts.list
+ *
+ * @example
+ *
+ * inquirer.registerPrompt('dynamicList', DynamicList);
+ *
+ * inquirer.prompt([{
+ *   message: ''What size do you need?'',
+ *   type: 'dynamicList',
+ *   choices: ['Jumbo', 'Large', 'Standard', 'Medium', 'Small', 'Micro'],
```
Comment (jviotti) [inline lib/cli/dynamic-list.js]: Can we add spaces between the items and the brackets for consistency with the remaining of the codebase?

```js
[ 'Jumbo', 'Large', 'Standard', 'Medium', 'Small', 'Micro' ]
```

File: lib/cli/drive-scanner.js [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+    })
+    .pausable(new Rx.Subject());
+};
+
+const emitChangeEvent = function(drives) {
+  const prevDrives = prevPollDrives.get(this);
+  if (prevDrives !== null && _.isEqual(drives, prevPollDrives)) {
+    return;
+  }
+  this.emit('change', drives, prevDrives);
```
Comment (jviotti) [inline lib/cli/drive-scanner.js]: Can we make this function not rely on the current context?

File: lib/cli/dynamic-list.js [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+ *   emitter: scanner,  // Event emitter with events `add` and `remove` and methods `start` and `stop`
+ *   name: 'size',
+ *   mapFn: _.identity  // mapFn is used transform one choice from the emitter into a `list` choice
+ *  });
+ */
+class DynamicList extends inquirer.prompt.prompts.list {
+
+  /**
+   * This is called by `inquirer`
```
Comment (jviotti) [inline lib/cli/dynamic-list.js]: See https://github.com/resin-io/resin-cli-visuals/blob/master/lib/widgets/progress.coffee for a JavaScript class JSDoc annotation example. The annotation that contains `@class` should be actually on the constructor. The methods should contain `@method`, access types (e.g: `@public`), etc.

File: lib/cli/drive-scanner.js [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+
+    drivesStream.get(this).subscribe((drives) => {
+      drives.sort(driveCompareFn);
+      Reflect.apply(emitChangeEvent, this, [ drives ]);
+      prevPollDrives.set(this, drives);
+    });
+  }
+
+  /**
+   * @summary Start scanning drives.
```
Comment (jviotti) [inline lib/cli/drive-scanner.js]: What is the access type of these methods?

File: lib/cli/dynamic-list.js [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+ *
+ * inquirer.registerPrompt('dynamicList', DynamicList);
+ *
+ * inquirer.prompt([{
+ *   message: ''What size do you need?'',
+ *   type: 'dynamicList',
+ *   choices: ['Jumbo', 'Large', 'Standard', 'Medium', 'Small', 'Micro'],
+ *   emitter: scanner,  // Event emitter with events `add` and `remove` and methods `start` and `stop`
+ *   name: 'size',
+ *   mapFn: _.identity  // mapFn is used transform one choice from the emitter into a `list` choice
```
Comment (jviotti) [inline lib/cli/dynamic-list.js]: I think we're missing one piece in this design, which the actual "drive selection" widget. We have an inquirer "dynamic list" and a "drive scanner". The "drive selection" widget would be the one that ties them up, since `etcher.js` is currently dealing with the low-level details of that.

File: npm-shrinkwrap.json (hunk 2 above)
Comment (stefan-mihaila) [inline npm-shrinkwrap.json]: I thought I did, but... err... nevermind 🙃

**Comment by jviotti** [comment]
@lurch I agree. As a consequence, we need to have a function to return "labels" that can be re-used between the CLI and the GUI.

File: lib/cli/drive-scanner.js (hunk 6 above)
Comment (stefan-mihaila) [inline lib/cli/drive-scanner.js]: They're private instance variables. (Different instances of `DriveScanner` would have access to different values) My first try was to use private by convention (`this._interval`), but dangling underscores aren't permitted in `ESLint` so I assumed you don't want them.

**Comment by stefan-mihaila** [comment]
@lurch At the moment there is no label for fixed drives in GUI, is there? Should I remove it in the CLI?
screen shot 2016-12-14 at 18 36 05

**Comment by jviotti** [comment]
That is because we don't show system drives at all on the GUI, but we will in the future. See https://github.com/resin-io/etcher/issues/888. The label will be "SYSTEM DRIVE".

**Comment by stefan-mihaila** [comment]
@jviotti We show them with the "unsafe" or what's its name option enabled in settings.
What labels we want from the CLI in this PR?

**Comment by jviotti** [comment]
Let’s put in this PR the ones that currently appear on the GUI.

--
Juan Cruz Viotti
Software Engineer, Resin.io (https://resin.io)

File: lib/cli/drive-scanner.js (hunk 6 above)
Comment (stefan-mihaila) [inline lib/cli/drive-scanner.js]: What do you mean by only one key? Different instances of `DriveScanner` would use different keys in the map, won't them?

File: lib/cli/drive-scanner.js (hunk 13 above)
Comment (stefan-mihaila) [inline lib/cli/drive-scanner.js]: I can move it in the class. The reason I extracted it is because I thought it would make the constructor more readable. Once again, I created it having a private method in mind and it made sense to me to be here as one of the `DriveScanner`'s private methods, and it looked like it won't cause damage because this whole module exists just to accomodate the `DriveScanner`.
On the other hand, I admit that I don't know if that's the way to do things in JS or not, maybe it's an anti-pattern. I think I need to read some more on the subject of private instance methods and variables in JS best practices and patterns.

File: lib/cli/dynamic-list.js (hunk 8 above)
Comment (stefan-mihaila) [inline lib/cli/dynamic-list.js]: I'll get rid of it

File: lib/cli/etcher.js [hunk 17]
```diff
@@ -18,61 +18,115 @@
...
+    choice.disabled = 'TOO SMALL FOR IMAGE';
+  }
+  return choice;
+};
+
+const mustConfirm = !options.robot && !options.yes;
+
+Bluebird.all([
+  isElevated(),
+  imageStream.getImageMetadata(options._[0])
```
Comment (stefan-mihaila) [inline lib/cli/etcher.js]: Wasteful in what sense? My thought process was that, doing it like this would be faster (as in more responsive for the user) in the positive case.

Comment (jviotti) [inline lib/cli/etcher.js]: The `isElevated` function is really cheap to execute. Its code is just:

```
return process.getuid && process.getuid() === 0;
```

So the savings we get from parallelizing this are minimal, and if the user indeed lacks privileges, we have to wait for the `getImageMetadata()` function, which is way more expensive, just to be able to say to the user that he doesn't have permissions to proceed.

My way of thinking is that since `isElevated` is so cheap, doing `isElevated` and `.getImageMetadata()` in sequence is unnoticeable from doing them in parallel, and still provides a "fast failure" for the case where the user lacks the privileges.

**Comment by stefan-mihaila** [comment]
@jviotti I `git push --force` without realising it will screw up all comments in the added files. I tried to revert and push force again, but it looks like it didn't do the trick. At the moment not all your comments are addressed, I will notify you when they are so you can review again.

> Let’s put in this PR the ones that currently appear on the GUI.

Do you want me to merge some other branch here and take over?

File: lib/cli/etcher.js [hunk 18]
```diff
@@ -18,61 +18,115 @@
...
+const DynamicList = require('./dynamic-list');
+const DriveScanner = require('./drive-scanner');
 const EXIT_CODES = require('../src/exit-codes');
 
-isElevated().then((elevated) => {
+inquirer.registerPrompt('dynamicList', DynamicList);
+
+const scanner = new DriveScanner(2000);
+
+const driveToChoice = (drive, minSizeToEnable) => {
```
Comment (stefan-mihaila) [inline lib/cli/etcher.js]: Not sure if `drive-scanner` should care about image size either...

File: lib/cli/etcher.js [hunk 19]
```diff
@@ -18,61 +18,115 @@
...
+
+  if (drive.system) {
+    displayName += chalk.red(' FIXED DRIVE');
+  }
+
+  const choice = {
+    name: displayName,
+    value: drive.device
+  };
+  if (drive.size < minSizeToEnable) {
```
Comment (stefan-mihaila) [inline lib/cli/etcher.js]: It permits it actually, as it displays the disabled label iff drive size is strictly lesser than the image size.
Anyway, by reusing logic from the `SelectionStateModel` you meant actually importing it here? Or maybe refactoring some useful methods to a `common` module?

File: lib/cli/etcher.js [hunk 20]
```diff
@@ -18,61 +18,115 @@
...
+  return choice;
+};
+
+const mustConfirm = !options.robot && !options.yes;
+
+Bluebird.all([
+  isElevated(),
+  imageStream.getImageMetadata(options._[0])
+]).then((results) => {
+  const elevated = results[0];
```
Comment (stefan-mihaila) [inline lib/cli/etcher.js]: Nice one, thanks

File: lib/cli/etcher.js (hunk 4 above)
Comment (stefan-mihaila) [inline lib/cli/etcher.js]: Not sure what logic I can reuse for this condition.

**Comment by jviotti** [comment]
> @jviotti I git push --force without realising it will screw up all comments in the added files. I tried to revert and push force again, but it looks like it didn't do the trick. At the moment not all your comments are addressed, I will notify you when they are so you can review again.

You can still reveal the "outdated" comments from the GH UI:

screenshot 2016-12-16 11 26 39

> Do you want me to merge some other branch here and take over?

I mean lets put in the CLI all the labels that exist currently on the GUI.

File: lib/cli/etcher.js (hunk 4 above)
Comment (jviotti) [inline lib/cli/etcher.js]: Oops, my bad. `SelectionStateModel.isSystemDrive()` was currently in a branch. I got merged very recently (https://github.com/resin-io/etcher/pull/978), so you can rebase to get it.

File: lib/cli/etcher.js (hunk 19 above)
Comment (jviotti) [inline lib/cli/etcher.js]: Oh, I see, my bad.

> Anyway, by reusing logic from the SelectionStateModel you meant actually importing it here? Or maybe refactoring some useful methods to a common module?

It'd be really cool if we could re-use it as it is, however `SelectionStateModel` is an Angular service, and can't be used outside such framework. I propose extracting all the "static" functions that module contains (that don't interact with the application state), such as `isSystemDrive()`, `isDriveLocked()`, etc; as a separate NodeJS module we can also import in the CLI.

Some things to keep in mind:

- These functions contain tests, so the tests will have to be extracted as well.
- The extracted module will still need to be exposed as an Angular service so the Angular components that currently use these functions can still do so (you can collaborate with Shou on this one).

The current approach to re-usable code between the GUI and CLI is to put it on `lib/src` (I know the name is not ideal, we should probably rename to `lib/shared`), so the extracted functionality can live there, and then there can be another small module in `lib/gui` that simply requires that logic and re-exposes it as an Angular service.

In the future, we want to have a re-usable module that handles the "labels" we put to drives (such as system, locked, etc) and also be able to automatically compute a list of them (so we don't hardcode the label names anywhere), so that would be the perfect place to include that.

Let's make those changes as another PR, so its easier to review, and then you can rebase here.

File: lib/cli/etcher.js (hunk 18 above)
Comment (jviotti) [inline lib/cli/etcher.js]: Yeah, we should find a way to have all the logic necessary to display the drive options in the drive selection widget, while still have the knowledge about the image in the main file.

**Comment by lurch** [comment]
> all the labels that exist currently on the GUI.

Which, to be fair, is currently a moving target...

**Comment by jviotti** [comment]
Yeah, let's try our best to keep this updated. I expect this PR to be merged soon, so that shouldn't be a major problem.

**Comment by stefan-mihaila** [comment]
@jviotti I've update the PR. One observation is that currently, with "Unsafe mode" enabled, the GUI doesn't prompt the user for any confirmation when flashing an image to a "SYSTEM DRIVE" (non removable actually). It looked to me to dangerous to not prompt the user for confirmation in the CLI (at least in the GUI the user has to activate "unsafe mode", there is no such thing in the CLI)

File: lib/cli/drive-selection-widget.js [hunk 21]
```diff
@@ -0,0 +1,140 @@
...
+const driveToChoiceFn = (image) => {
+
+  /**
+   * Map a drive object to a `inquirer` choice object
+   * @param {object} drive - Drive object as returned by `drivelist`
+   * @private
+   * @returns {{name: string, value: object}}
+   */
+  const driveToChoice = (drive) => {
+    const size = drive.size / 1000000000;
```
Comment (jviotti) [inline lib/cli/drive-selection-widget.js]: We have this exact same logic in `lib/gui/utils/byte-size/filters/gigabyte.js`. Maybe worth moving to `lib/shared`? :)

File: lib/cli/drive-selection-widget.js [hunk 22]
```diff
@@ -0,0 +1,140 @@
+/*
```
Comment (jviotti) [inline lib/cli/drive-selection-widget.js]: I'm not going to comment on each one, but the JSDoc should contain their title preceded by `@summary`, the types should be capitalized (e.g: `Number`, `Object`), every annotation should contain the access type, and the type of annotation (e.g: `@function`), and each function should contain an `@example`.

File: lib/cli/drive-selection-widget.js [hunk 23]
```diff
@@ -0,0 +1,140 @@
...
+
+    const driveInfo = `${drive.device} (${size.toFixed(1)} GB) - ${drive.description}`;
+
+    const systemDriveText = chalk.red(constraints.labels.SYSTEM_DRIVE);
+    const systemDriveDisplay = constraints.isSystemDrive(drive) ? systemDriveText : '';
+
+    const notRecommendedSizeText = chalk.yellow(constraints.labels.NOT_RECOMMENDED);
+    const recommendedSizeDisplay = constraints.isDriveSizeRecommended(drive, image) ? '' : notRecommendedSizeText;
+
+    const displayName = `${driveInfo} ${systemDriveDisplay} ${recommendedSizeDisplay}`;
```
Comment (jviotti) [inline lib/cli/drive-selection-widget.js]: If the drive is not a system drive `systemDriveDisplay` will be an empty string. If the drive size is not recommended, this will result in the following string (notice the double spaces after the drive description):

```
/dev/disk2 (2 GB) - Foo Bar  NOT RECOMMENDED
```

Comment (jviotti) [inline lib/cli/drive-selection-widget.js]: Same for the converse scenario. If the drive is a system one and the size is recommended, there will be trailing spaces.

File: lib/cli/drive-selection-widget.js [hunk 24]
```diff
@@ -0,0 +1,140 @@
...
+
+    const choice = {
+      name: displayName,
+      value: drive
+    };
+
+    if (!constraints.isDriveLargeEnough(drive, image)) {
+      choice.disabled = constraints.labels.TOO_SMALL;
```
Comment (jviotti) [inline lib/cli/drive-selection-widget.js]: Does the `disabled` property have any meaning? Why is it assigned to a string? I can't find it on the docs.

File: lib/cli/drive-selection-widget.js [hunk 25]
```diff
@@ -0,0 +1,140 @@
...
+  const confirmWriteToNotRecommendedDrive = (answers) => {
+    if (!confirmationRequired || !answers.yes) {
+      return false;
+    }
+    return !constraints.isDriveSizeRecommended(answers.drive, image);
+  };
+
+  initialDrives.sort(driveComparer);
+
+  return inquirer.prompt([ {
```
Comment (jviotti) [inline lib/cli/drive-selection-widget.js]: Let's open the object in a separate line.

File: lib/cli/drive-selection-widget.js [hunk 26]
```diff
@@ -0,0 +1,140 @@
...
+    type: 'dynamicList',
+    name: 'drive',
+    choices: initialDrives.map(driveToChoiceFn(image)),
+    emitter: scanner,
+    mapFn: driveToChoiceFn(image),
+    sort: true,
+    compareFn: (choice1, choice2) => {
+      return driveComparer(choice1.value, choice2.value);
+    }
+  }, {
```
Comment (jviotti) [inline lib/cli/drive-selection-widget.js]: Let's open the objects in separate lines here and below as well.

File: lib/cli/drive-selection-widget.js [hunk 27]
```diff
@@ -0,0 +1,140 @@
...
+    when: confirmWriteToFixedDrive
+  }, {
+    message: (answers) => {
+      return constraints.labels.unrecommendedSizeWarningText(answers.drive, image);
+    },
+    type: 'confirm',
+    name: 'yes',
+    default: false,
+    when: confirmWriteToNotRecommendedDrive
+  } ]).then((answers) => {
```
Comment (jviotti) [inline lib/cli/drive-selection-widget.js]: The closing `]` should be on a separate line.

File: lib/cli/drive-selection-widget.js [hunk 28]
```diff
@@ -0,0 +1,140 @@
...
+      return driveComparer(choice1.value, choice2.value);
+    }
+  }, {
+    message: 'This will erase the selected drive. Are you sure?',
+    type: 'confirm',
+    name: 'yes',
+    default: false,
+    when: _.constant(confirmationRequired)
+  }, {
+    message: `You have selected a ${chalk.red('NON REMOVABLE DRIVE')}. Do you want to continue?`,
```
Comment (jviotti) [inline lib/cli/drive-selection-widget.js]: I wonder if we should re-use the `SYSTEM_DRIVE` label here.

File: lib/cli/drive-selection-widget.js [hunk 29]
```diff
@@ -0,0 +1,140 @@
...
+  /**
+   * Map a drive object to a `inquirer` choice object
+   * @param {object} drive - Drive object as returned by `drivelist`
+   * @private
+   * @returns {{name: string, value: object}}
+   */
+  const driveToChoice = (drive) => {
+    const size = drive.size / 1000000000;
+
+    const driveInfo = `${drive.device} (${size.toFixed(1)} GB) - ${drive.description}`;
```
Comment (jviotti) [inline lib/cli/drive-selection-widget.js]: Can we extract all the logic to create a descriptive line for a drive (including labels, etc) to a private exported function in this file, so we can unit test it?

File: lib/cli/drive-selection-widget.js [hunk 30]
```diff
@@ -0,0 +1,140 @@
...
+const _ = require('lodash');
+
+const DynamicList = require('./dynamic-list');
+const scanner = require('./drive-scanner');
+const driveComparer = require('../shared/utils').driveComparer;
+const constraints = require('../shared/drive-constraints');
+
+inquirer.registerPrompt('dynamicList', DynamicList);
+
+const driveToChoiceFn = (image) => {
```
Comment (jviotti) [inline lib/cli/drive-selection-widget.js]: This function should contain a JSDoc annotation. Also, let's drop the `Fn` prefix.

File: lib/cli/drive-selection-widget.js [hunk 31]
```diff
@@ -0,0 +1,140 @@
...
+
+  /**
+   * Map a drive object to a `inquirer` choice object
+   * @param {object} drive - Drive object as returned by `drivelist`
+   * @private
+   * @returns {{name: string, value: object}}
```
Comment (jviotti) [inline lib/cli/drive-selection-widget.js]: I don't think this is valid JSDoc. You can say `@returns {Object} an inquirer choice object`. Users wanting to know more about what that means can look at Inquirer's documentation, at the function example, or at the unit tests.

File: lib/cli/drive-selection-widget.js [hunk 32]
```diff
@@ -0,0 +1,140 @@
...
+      choice.disabled = constraints.labels.LOCKED;
+    }
+
+    return choice;
+  };
+
+  return driveToChoice;
+};
+
+module.exports = function(initialDrives, confirmationRequired, image) {
```
Comment (jviotti) [inline lib/cli/drive-selection-widget.js]: Can we not export the function directly, but export it as property, like `exports.display()`. The benefit of the latter case is that we can stub this function on the client's code if needed, while a plain function can't be stubbed.

Comment (jviotti) [inline lib/cli/drive-selection-widget.js]: Do you think the `initialDrives` is really required? As far as I can see, the drive scanner will scan right away, so we can safely delay showing the list until we receive the first set of data (which should be just a couple of milliseconds anyway).

Comment (jviotti) [inline lib/cli/drive-selection-widget.js]: Can we transform this function signature to take an object instead? For example:

```
{
    image: { ... },
    omitConfirmation: true 
}
```

Comment (jviotti) [inline lib/cli/drive-selection-widget.js]: Also, I think its nice to change `confirmationRequired` to the opposite meaning: e.g: `omitConfirmation`. This has the benefit that confirmation is the default (which is a great default), and that you don't have to negate `options.yes` and `robot.isEnabled()` on the client code.

File: lib/cli/drive-selection-widget.js [hunk 33]
```diff
@@ -0,0 +1,140 @@
...
+
+  /**
+   * Returns true iff `inquirer` should prompt in the case of a fixed drive selected
+   * @param {object} answers - The answers given so far
+   * @returns {boolean}
+   */
+  const confirmWriteToFixedDrive = (answers) => {
+    if (!confirmationRequired || !answers.yes) {
```
Comment (jviotti) [inline lib/cli/drive-selection-widget.js]: This snippet is exactly the same in the function below. Maybe we can extract and re-use? Also, if this piece of code is extracted, the logic would be so simple that I think we could inline it on the inquirer questions below.

**Comment by jviotti** [comment]
> One observation is that currently, with "Unsafe mode" enabled, the GUI doesn't prompt the user for any confirmation when flashing an image to a "SYSTEM DRIVE" (non removable actually). It looked to me to dangerous to not prompt the user for confirmation in the CLI (at least in the GUI the user has to activate "unsafe mode", there is no such thing in the CLI)

Sounds totally reasonable! :+1:

File: lib/cli/drive-selection-widget.js (hunk 29 above)
Comment (jviotti) [inline lib/cli/drive-selection-widget.js]: Actually, this WHOLE logic would be re-used between the CLI and the GUI, and not just the byte to gigabyte function I mentioned above. The GUI displays the exact same thing, but it creates separate HTML elements for the labels.

We can build a generic function that returns an object that looks like this:

```javascript
{
   description: '/dev/disk2 (2 GB) - Foo bar',
   labels: [
      'SYSTEM DRIVE',
      'LOCKED'
   ]
}
```

Then, in the GUI, we can loop between the labels in Angular and create elements, while in the CLI, we can join the label values and append it to the description.

**Comment by jviotti** [comment]
@stefan-mihaila I spotted various pieces of functionality that could be moved to `lib/shared` in other to be re-used between the CLI and the GUI. I think its better to handle those as separate PRs and rebase in order to not grow this into a monster PR :)

**Comment by jviotti** [comment]
I might help you with some of the GUI/CLI `lib/shared` in the mid-time. I'll let you know if I make any of those myself!

File: lib/cli/drive-selection-widget.js (hunk 21 above)
Comment (jviotti) [inline lib/cli/drive-selection-widget.js]: See https://github.com/resin-io/etcher/pull/1021

File: lib/shared/drive-constraints.js [hunk 34]
```diff
@@ -147,3 +147,22 @@ exports.isDriveValid = (drive, image) => {
...
+ * @summary Constant strings used as drive labels
+ * @object
+ * @public
+ */
+exports.labels = {
+  TOO_SMALL: 'TOO SMALL FOR IMAGE',
+  LOCKED: 'LOCKED',
+  SYSTEM_DRIVE: 'SYSTEM DRIVE',
+  NOT_RECOMMENDED: 'NOT RECOMMENDED',
+  unrecommendedSizeWarningText: (drive, image) => {
```
Comment (jviotti) [inline lib/shared/drive-constraints.js]: See https://github.com/resin-io/etcher/pull/1022

File: lib/cli/drive-scanner.js [hunk 35]
```diff
@@ -0,0 +1,129 @@
...
+ *   console.log(currentDrives , previousDrives);
+ * }
+ *
+ * // Start emitting events
+ * scanner.start()
+ *
+ * // Stop emitting events
+ * scanner.stop()
+ */
+module.exports = Object.create(new EventEmitter(), {
```
Comment (stefan-mihaila) [inline lib/cli/drive-scanner.js]: `class DriveEmitter extends EventEmitter` was my first choice, but I changed this to be a singleton (https://github.com/resin-io/etcher/pull/976#discussion_r92642941). It's true that I could have made it a singleton just outside this module (as in creating a `DriveEmitter` class and only exporting an instance), but since having multiple instances can be problematic, it seemed right to me to implement it like this.
Regarding composition, looking at [lib/gui/modules/drive-scanner.js](https://github.com/resin-io/etcher/blob/6456a8fe967ffe52f089828c158320d9389cc34d/lib/gui/modules/drive-scanner.js#L114-L114), it uses a nice trick because it returns the emitter. So that's not implemented as a singleton, it's just a function constructor that returns the emitter instead of the implicit `this`, but AngularJS uses it as a singleton (I mean it only creates one instance and uses that instance wherever it's injected). Without this trick, with composition I would also have to use some adaptor methods.

File: lib/cli/drive-selection-widget.js (hunk 24 above)
Comment (stefan-mihaila) [inline lib/cli/drive-selection-widget.js]: Yes, fortunately it works for lists as well as for choices. You can see it used in the `Inquirer.js` examples: https://github.com/SBoudrias/Inquirer.js/blob/master/examples/list.js#L20

File: lib/cli/drive-scanner.js [hunk 36]
```diff
@@ -0,0 +1,129 @@
...
+  stop: {
+    value: () => {
+      previousDrives = null;
+      drivesStream.pause();
+    }
+  }
+});
+
+drivesStream.subscribe((drives) => {
+  if (previousDrives !== null && _.isEqual(drives, previousDrives)) {
```
Comment (stefan-mihaila) [inline lib/cli/drive-scanner.js]: you're right

File: lib/cli/drive-scanner.js [hunk 37]
```diff
@@ -0,0 +1,129 @@
...
+    return;
+  }
+
+  module.exports.emit('change', drives, previousDrives);
```
Comment (stefan-mihaila) [inline lib/cli/drive-scanner.js]: yes, we don't use it as an event, so I won't emit it

File: tests/cli/drive-scanner.spec.js [hunk 38]
```diff
@@ -0,0 +1,112 @@
...
+
+const m = require('mochainon');
+const drivelist = require('drivelist');
+
+describe('CLI: DriveScanner', function() {
+
+  describe('given no available drives', function() {
+
+    beforeEach(function() {
+      this.scanner = require('../../lib/cli/drive-scanner');
```
Comment (stefan-mihaila) [inline tests/cli/drive-scanner.spec.js]: Yes, it doesn't help with anything since it's a singleton anyway.
This brings up a concern. Do tests run in parallel? If I have a test that needs the scanner to not be started to function correctly, and another test starts the scanner, since the scanner is a singleton, will the tests interfere?

Comment (stefan-mihaila) [inline tests/cli/drive-scanner.spec.js]: Just realised... This can be a problem in code, not in tests. I mean, having a singleton with `start` and `stop` is not a good idea, because it will interact everywhere... Looking into it...

Comment (jviotti) [inline tests/cli/drive-scanner.spec.js]: The reason for a singleton scanner is that having multiple scanners running at the same time can be *very* troubling, specifically for Windows, which will start throwing cryptic sporadic issues, so a singleton was my way to prevent that from happing at all.

The tests don't run in parallel, so it should be fine. I think that we can keep it as a singleton as long as we have some reliable way to reset the state (if any). Looks like that could be part of `.stop()`.

File: lib/cli/drive-selection-widget.js (hunk 24 above)
Comment (jviotti) [inline lib/cli/drive-selection-widget.js]: Oh, I see. So its some kind of informative message you show to explain why an element is disabled, but it doesn't control whether the option is disabled or not, right?

If so, I think it might be good to add a comment explaining it, since it looks really awkward at first sight (definitely not your fault. The interface exposed by inquirer for this is very weird).

File: lib/cli/drive-scanner.js (hunk 35 above)
Comment (jviotti) [inline lib/cli/drive-scanner.js]: > class DriveEmitter extends EventEmitter was my first choice, but I changed this to be a singleton (#976 (comment)). It's true that I could have made it a singleton just outside this module (as in creating a DriveEmitter class and only exporting an instance), but since having multiple instances can be problematic, it seemed right to me to implement it like this.

My thought was that we could use `class DriveEmitter extends EventEmitter` to declare it, but only export `new DriveEmitter()`.

> Regarding composition, looking at lib/gui/modules/drive-scanner.js, it uses a nice trick because it returns the emitter. So that's not implemented as a singleton, it's just a function constructor that returns the emitter instead of the implicit this, but AngularJS uses it as a singleton (I mean it only creates one instance and uses that instance wherever it's injected). Without this trick, with composition I would also have to write some adaptor methods.

NodeJS will not parse a file twice. That means that if you `require()` the same file from multiple modules, they will all get the same instance of it so it should just work, unless I'm missing something?

File: lib/cli/drive-selection-widget.js (hunk 24 above)
Comment (stefan-mihaila) [inline lib/cli/drive-selection-widget.js]: It controls if the option is disabled or not and also displays the informative string. Yes, you're right, it looks awkward, I'll add a comment.

File: tests/cli/drive-scanner.spec.js (hunk 38 above)
Comment (stefan-mihaila) [inline tests/cli/drive-scanner.spec.js]: OK, I'm not worried anymore about this in tests, what I'm worried about is that the users of the scanner (us) can't rely on a started scanner to not be stopped in some other part of the application. I'm not saying that it shouldn't be a singleton, just that there's an annoyance with it being one and I was wandering if there's a good way to get around it.

Comment (jviotti) [inline tests/cli/drive-scanner.spec.js]: Hm I see what you mean. We could either:

- Be careful about our own use with code reviews, etc (error prone)
- Provide some sort of mechanism to prevent this from happening

Any thoughts about the second option? Maybe some sort of lock?
